File: packages/kimi-agent/src/kosong/provider/providers/standard.rs
/// Standard provider definitions — the four canonical vendors.
///
/// Corresponds to `kosong/provider/providers/standard.contrib.ts`.
use crate::kosong::protocol::protocol::Protocol;
use crate::kosong::protocol::protocol_trait::{ProtocolEndpoint, ProtocolTrait, TraitContext};
use crate::kosong::provider::provider_definition::{register_provider_definition, ProviderDefinition};

/// Register the standard provider definitions.
pub fn register() {
    // Anthropic
    register_provider_definition(ProviderDefinition {
        id: "anthropic".to_string(),
        base_protocol: Protocol::Anthropic,
        traits: vec![],
        endpoint: Some(ProtocolEndpoint {
            api_key_env: Some("ANTHROPIC_API_KEY".to_string()),
            base_url_env: Some("ANTHROPIC_BASE_URL".to_string()),
            default_base_url: None,
        }),
        host_headers: None,
        model_source: None,
    });

    // OpenAI
    register_provider_definition(ProviderDefinition {
        id: "openai".to_string(),
        base_protocol: Protocol::OpenAI,
        traits: vec![],
        endpoint: Some(ProtocolEndpoint {
            api_key_env: Some("OPENAI_API_KEY".to_string()),
            base_url_env: Some("OPENAI_BASE_URL".to_string()),
            default_base_url: None,
        }),
        host_headers: None,
        model_source: None,
    });

    // OpenAI Responses
    register_provider_definition(ProviderDefinition {
        id: "openai_responses".to_string(),
        base_protocol: Protocol::OpenaiResponses,
        traits: vec![],
        endpoint: Some(ProtocolEndpoint {
            api_key_env: Some("OPENAI_API_KEY".to_string()),
            base_url_env: Some("OPENAI_BASE_URL".to_string()),
            default_base_url: None,
        }),
        host_headers: None,
        model_source: None,
    });

    // Google GenAI
    register_provider_definition(ProviderDefinition {
        id: "google-genai".to_string(),
        base_protocol: Protocol::GoogleGenAI,
        endpoint: None,
        traits: vec![
            // Vertex AI fallback chain
            ProtocolTrait {
                strict_thinking_validation: false,
                provides: None,
                endpoint: Some(|_: &TraitContext| -> Option<ProtocolEndpoint> {
                    Some(ProtocolEndpoint {
                        api_key_env: Some("VERTEXAI_API_KEY".to_string()),
                        base_url_env: Some("GOOGLE_VERTEX_BASE_URL".to_string()),
                        default_base_url: None,
                    })
                }),
                default_headers: None,
                convert_tool: None,
                convert_message: None,
                merge_history: None,
                build_params: None,
                tool_call_id_policy: None,
                with_thinking: None,
                preserve_thinking: None,
                with_max_completion_tokens: None,
                cache_key: None,
                extract_usage: None,
                reasoning_key: None,
                capability: None,
                upload_video: None,
            },
            // Plain Gemini fallback
            ProtocolTrait {
                strict_thinking_validation: false,
                provides: None,
                endpoint: Some(|_: &TraitContext| -> Option<ProtocolEndpoint> {
                    Some(ProtocolEndpoint {
                        api_key_env: Some("GOOGLE_API_KEY".to_string()),
                        base_url_env: Some("GOOGLE_GEMINI_BASE_URL".to_string()),
                        default_base_url: None,
                    })
                }),
                default_headers: None,
                convert_tool: None,
                convert_message: None,
                merge_history: None,
                build_params: None,
                tool_call_id_policy: None,
                with_thinking: None,
                preserve_thinking: None,
                with_max_completion_tokens: None,
                cache_key: None,
                extract_usage: None,
                reasoning_key: None,
                capability: None,
                upload_video: None,
            },
        ],
        host_headers: None,
        model_source: None,
    });
}

File: packages/agent-core-v2/src/human/llm/provider/providers/standard.ts
import type { ProviderConnection } from '#/llm/protocol/connection';
import { createProvider } from '#/llm/provider/definition';
import { anthropicBase } from '#/llm/requester/bases/anthropic/requester';
import { createGoogleGenAIBase, googleGenAIBase } from '#/llm/requester/bases/google-genai/requester';
import { openAIBase } from '#/llm/requester/bases/openai/requester';
import { openAIResponsesBase } from '#/llm/requester/bases/openai-responses/requester';

export const openAIConnection: ProviderConnection = {
  endpoint: () => ({ apiKeyEnv: 'OPENAI_API_KEY', baseUrlEnv: 'OPENAI_BASE_URL' }),
};

export const anthropicConnection: ProviderConnection = {
  endpoint: () => ({ apiKeyEnv: 'ANTHROPIC_API_KEY', baseUrlEnv: 'ANTHROPIC_BASE_URL' }),
};

export const geminiConnection: ProviderConnection = {
  endpoint: () => ({ apiKeyEnv: 'GOOGLE_API_KEY', baseUrlEnv: 'GOOGLE_GEMINI_BASE_URL' }),
};

export const vertexConnection: ProviderConnection = {
  endpoint: () => ({ apiKeyEnv: 'VERTEXAI_API_KEY', baseUrlEnv: 'GOOGLE_VERTEX_BASE_URL' }),
};

export const openaiProvider = createProvider({
  id: 'openai',
  protocols: {
    openai: { base: openAIBase, connection: openAIConnection },
    openai_responses: { base: openAIResponsesBase, connection: openAIConnection },
  },
});

export const anthropicProvider = createProvider({
  id: 'anthropic',
  protocols: {
    anthropic: { base: anthropicBase, connection: anthropicConnection },
  },
});

export const googleProvider = createProvider({
  id: 'google',
  protocols: {
    'google-genai': { base: googleGenAIBase, connection: geminiConnection },
    'google-vertex': {
      base: createGoogleGenAIBase({ vertexai: true }),
      connection: vertexConnection,
    },
  },
  media: { inlineVideo: true },
});
